materialize-bigquery/schema.go

```diff
@@ -0,0 +1,292 @@
+		case "string":
+			if projection.Inference.String_ != nil {
+				s := projection.Inference.String_
+				switch s.Format {
+				case "date-time":
+					f.fieldType = bigquery.TimestampFieldType
+				case "date":
+					f.fieldType = bigquery.DateFieldType
+				// There is a case where a String_ is present but the values aren't set.
+				// I'm not sure how this can happen since the protobuf struct should
+				// be optional and nil, but it did happen in my testing and I couldn't
+				// figured out why, so this is here as a precaution (P-O)
+				case "":
+					f.fieldType = bigquery.StringFieldType
+					f.sanitizers = append(f.sanitizers, sqlDriver.DefaultQuoteSanitizer)
+
+				default:
+					return nil, fmt.Errorf("inferring string column type as the format is not supported: %s", s.Format)
+				}
+			} else {
+				f.fieldType = bigquery.StringFieldType
+				f.sanitizers = append(f.sanitizers, sqlDriver.DefaultQuoteSanitizer)
+			}
+		case "boolean":
+			f.fieldType = bigquery.BooleanFieldType
+		case "integer":
+			f.fieldType = bigquery.IntegerFieldType
+		case "number":
+			f.fieldType = bigquery.BigNumericFieldType
+		case "object":
+			f.fieldType = bigquery.StringFieldType
+			f.sanitizers = append(f.sanitizers, sqlDriver.DefaultQuoteSanitizer, sqlDriver.SingleQuotesWrapper())
+		case "array":
+			f.fieldType = bigquery.StringFieldType
+			f.sanitizers = append(f.sanitizers, sqlDriver.DefaultQuoteSanitizer, sqlDriver.SingleQuotesWrapper())
+		}
+	}
+
+	if f.fieldType == "" {
+		return nil, fmt.Errorf("could not map the field to a big query type: %s", projection.Field)
+	}
+
+	return f, nil
+}
+
+func (f *Field) FieldType() bigquery.FieldType {
+	return bigquery.FieldType(f.fieldType)
+}
+
+func (f *Field) Name() string {
+	return f.name
+}
+
+func (f *Field) Render(val interface{}) (str string, err error) {
+	str = fmt.Sprint(val)
+
+	for _, sanitizer := range f.sanitizers {
+		str = sanitizer(str)
+	}
+
+	return str, err
+}
+
+func (f *Field) FieldSchema(projection *pf.Projection) (*bigquery.FieldSchema, error) {
+	var includesNull bool
+
+	// This is a test to make sure that their is only 1 infered type for this projection
+	// ie. Inference.Types = [string, integer] is wrong. [string, null] is OK. [string, integer, null] is wrong
+	if !projection.Inference.IsSingleType() {
+		return nil, fmt.Errorf("don't expect multiple types besides Null and a real type, can't proceed: %s", &projection.Inference.Types)
+	}
+
+	for _, p := range projection.Inference.Types {
+		if p == "null" {
+			includesNull = true
+		}
+	}
+
+	return &bigquery.FieldSchema{
+		Name:        f.name,
+		Type:        f.FieldType(),
+		Required:    projection.Inference.Exists == pf.Inference_MUST && !includesNull, // If Required is False, it means it's nullable
+		Description: projection.Inference.Description,
+	}, nil
+}
+
+var identifierSanitizerRegexp = regexp.MustCompile(`[^\-\._0-9a-zA-Z]`)
+
+func identifierSanitizer(text string) string {
```

Comment:
I think we want to quote identifiers that contain those characters instead of modifying them. It's less surprising for users if their column names show up exactly as they gave them. And if we replace characters with `_` then we could end up with duplicate column names since, for example, `foo/bar` and `foo$bar` would end up as `foo_bar`.

Reply:
This piece of code was carried over from the past implementation, if I make the change to what you suggest, it will be a breaking change from the previous implementation. We're good with it?